core/server/proxy/src/main/java/alluxio/proxy/s3/tag/TagHelper.java

```diff
@@ -0,0 +1,120 @@
+import java.io.InputStream;
+import java.io.StringWriter;
+import java.nio.charset.Charset;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Helper class for handling tags according to https://docs.aws.amazon.com/AmazonS3/latest/API/API_GetBucketTagging.html
+ * and https://docs.aws.amazon.com/AmazonS3/latest/API/API_GetObjectTagging.html.
+ */
+public class TagHelper {
+  private static final Logger LOG = LoggerFactory.getLogger(S3RestServiceHandler.class);
+
+  /**
+   * all tag information will go into TAG_FOLDER (/_TAG), inside of the folder, every file in the
+   * rest of the directory structure will have a corresponding file in the location inside the
+   * TAG_FOLDER if there are tags. Folders will have a reserved file name FOLDER_OBJECT_TAG_FILE
+   * which will represent the tags for that folder object.
+   */
+  private static final String TAG_FOLDER = "/_TAG";
+  private static final String FOLDER_OBJECT_TAG_FILE = "_FOLDER_TAG";
+
+  /**
+   * @param fs the filesystem
+   * @param path the filesystem path to update tags
+   * @param is the input stream of XML containing new tags
+   */
+  public static void putTags(FileSystem fs, String path, InputStream is)
+      throws IOException, AlluxioException {
+    final String tagPath = getTagPath(fs, path);
+
+    Map<String, String> tagMap = new HashMap<>();
+
+    final AlluxioURI tagPathURI = new AlluxioURI(tagPath);
+    LOG.info(tagPathURI.getPath());
+    try {
+      fs.delete(tagPathURI);
+    } catch (FileDoesNotExistException e) {
+      // ignore
+    }
+
+    StringWriter writer = new StringWriter();
+    IOUtils.copy(is, writer, Charset.defaultCharset());
+    String theString = writer.toString();
+
+    LOG.info(theString);
+
+    final Tagging tagToUpdate = new XmlMapper().readValue(theString, Tagging.class);
+
+    final TagSet tagSet = tagToUpdate.getTagSet();
+    if (tagSet != null) {
+      for (Tag tag : tagSet.getTags()) {
+        tagMap.put(tag.getKey(), tag.getValue());
+      }
+    }
+
+    try (final FileOutStream os = fs.createFile(tagPathURI,
+        CreateFilePOptions.newBuilder().setRecursive(true).build())) {
+      new ObjectMapper().writeValue(os, tagMap);
+    }
+  }
+
+  /**
+   * @param fs the file system
+   * @param path the path in the file system to look up tags
+   * @return map of tag keys to tag values
+   */
+  public static Map<String, String> getTags(FileSystem fs, String path)
+      throws IOException, AlluxioException {
+    final String tagPath = getTagPath(fs, path);
+
+    try {
+      final FileInStream fis = fs.openFile(new AlluxioURI(tagPath));
+      return new ObjectMapper().readValue(fis, Map.class);
+    } catch (FileDoesNotExistException e) {
+      return Collections.emptyMap();
+    }
+  }
+
+  private static String getTagPath(FileSystem fs, String path)
+      throws IOException, AlluxioException {
+    final URIStatus status = fs.getStatus(new AlluxioURI(path));
```

Comment:
Will this expensive to perform an RPC to convert paths?

Reply:
This is certainly going to be cheaper than opening a file which will need to happen later.